Add gce disk labels support via storageclass/CreateVolumeRequest parameters

ernestas2k opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds support to assign gce disk labels via storageclass / CreateVolumeRequest parameters.
 
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #340

**Special notes for your reviewer**:
- Any test I shouldn't have altered or should additionally update?
- note: API contract has changed - support of new "labels" parameter added.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Added support to assign gce disk labels by providing storageclass parameter `labels`.
```

ernestas2k commented:
/assign @mattcary

msau42 commented:
/ok-to-test

mattcary commented on README.md:
```diff
@@ -55,6 +55,7 @@ See Github [Issues](https://github.com/kubernetes-sigs/gcp-compute-persistent-di
 |------------------|---------------------------|---------------|----------------------------------------------------------------------------------------------------|
 | type             | `pd-ssd` OR `pd-standard` | `pd-standard` | Type allows you to choose between standard Persistent Disks  or Solid State Drive Persistent Disks |
 | replication-type | `none` OR `regional-pd`   | `none`        | Replication type allows you to choose between Zonal Persistent Disks or Regional Persistent Disks  |
+| labels           | `key1=value1,key2=value2` |               | Labels allow you to assign custom GCE Disk labels                                                  |
```
(dropping some notes here for future reference)

naming looks good, 'labels' is consistent with GCP label APIs:
https://cloud.google.com/sdk/gcloud/reference/compute/disks/add-labels
https://cloud.google.com/resource-manager/docs/creating-managing-labels

mattcary commented on examples/kubernetes/demo-regional-restricted-sc.yaml:
```diff
@@ -6,6 +6,7 @@ provisioner: pd.csi.storage.gke.io
 parameters:
   type: pd-standard
   replication-type: regional-pd
+  labels: key1=value1,key2=value2
```
Is it necessary to add labels to all the sample storageclasses if we already have demo-labeled-sc.yaml?

Given the amount these examples are copied/pasted I fear having them everywhere would lead to a lot of key1=value1 labels out in the world which disturbs me aesthetically :-P

mattcary commented on pkg/common/utils.go:
```diff
@@ -148,3 +148,36 @@ func CreateNodeID(project, zone, name string) string {
+	keyValueStrings := strings.Split(labels, labelsDelimiter)
+	for _, keyValue := range keyValueStrings {
+		keyValue := strings.Split(keyValue, labelsKeyValueDelimiter)
+
```
Do we want to validate the characters in the key/values? It would be nice to get immediate errors for this sort of thing rather than waiting for a possibly obscured error from a GCP response. While generally I don't think we want to duplicate GCP validation logic, doing some reasonable sanity checks might make sense.

eg from https://cloud.google.com/resource-manager/docs/creating-managing-labels:

Keys have a minimum length of 1 character and a maximum length of 63 characters, and cannot be empty. Values can be empty, and have a maximum length of 63 characters.
Keys and values can contain only lowercase letters, numeric characters, underscores, and dashes. All characters must use UTF-8 encoding, and international characters are allowed.
The key portion of a label must be unique. However, you can use the same key with multiple resources.
Keys must start with a lowercase letter or international character.

mattcary commented on pkg/common/utils.go:
```diff
@@ -148,3 +148,36 @@ func CreateNodeID(project, zone, name string) string {
+		if key == "" {
+			return nil, errIncorrectFormat
+		}
+
```
Are empty values ok?

ernestas2k replied:
Yes, according to the link you shared above:

_Keys have a minimum length of 1 character and a maximum length of 63 characters, and cannot be empty. Values can be empty, and have a maximum length of 63 characters._

mattcary replied:
Well at least I found the link :-P

Thanks

mattcary commented on pkg/gce-pd-csi-driver/controller_test.go:
```diff
@@ -668,6 +668,31 @@ func TestCreateVolumeArguments(t *testing.T) {
+			},
+		},
+		{
+			name: "success with malformed labels parameter",
```
Shouldn't this test be named "fail" and not "success"?

ernestas2k commented:
Thanks for fast feedback. I agree with all of the observations and I'll prepare updates.

Also, is there any integration test that I should update or actually introduce for this feature?

mattcary commented:
> Regarding the validation, absolutely let's do it, with slight caveat that changes on google side will have to be reflected in the code too (hopefully labels api won't change often).
> 
Great, thanks.


> Also, is there any integration test that I should update or actually introduce for this feature?
> 
That's a great question. The storage class test/k8s-integrations/configs/sc-standard.yaml is what gets used for the integration tests, adding a label there would be a good idea.


> The kubernetes-integration and verify jobs failed:
> verify: `E0729 17:31:41.945] Build timed out`

It succeeded after your last push.  You can run it manually with hack/verify-all.sh.

> integration: `W0729 17:52:03.603] - The zone 'projects/k8s-boskos-gce-project-07/zones/us-central1-b' does not have enough resources available to fulfill the request. Try a different zone, or try again later.`

Trying again later :-)


> 
> Is there anything to be done from my side regarding these failures?
> 
> PS: just pushed newest changes to address the comments (please take a look if you are ok with regexps, error messages, etc). I'm ready to rework it so it's not too different from other functionality.

Will look shortly, thanks!

ernestas2k commented:
Awesome, I just added e2e test case (altough I lean towards updating existing test case, such as `It("Should create and delete disk with default zone", func() {`). Let me know which of the options you prefer 👍

mattcary commented on pkg/common/utils.go:
```diff
@@ -148,3 +149,60 @@ func CreateNodeID(project, zone, name string) string {
+		return labelsMap, nil
+	}
+
+	regexKey, _ := regexp.Compile("^[a-z][-_a-z0-9]{0,62}$")
```
Use \p{Lowercase_Letter} instead of a-z? eg `^\p{Ll}[-_\p{Ll}0-9]{0,62}$`.

That way international characters are dealt with correctly, IIUC: https://www.regular-expressions.info/unicode.html

I tested it in go playground and it seems to work as expected.

mattcary commented:
> Awesome, I just added e2e test case `It("Should create and delete disk with labels", func() {`. To be honest, I lean more towards updating existing test case, such as `It("Should create and delete disk with default zone", func() {`.
> 
> What are your thoughts on this? 👍

I prefer adding a new test case, if it starts failing it will be a lot easier to tell what's going on; eg if the non-label version is still succeeding than we immediately know what's going on.

I think it would be good to add a label to the storage class in test/k8s-integeration/configs. This will be used for the k/k e2e tests and would be good to ferret out if something is depending on the storage class parameters in a way it shouldn't be. Also it will run on real PDs and give us, well, a real e2e test.

mattcary commented:
> I think it would be good to add a label to the storage class in test/k8s-integeration/configs. This will be used for the k/k e2e tests and would be good to ferret out if something is depending on the storage class parameters in a way it shouldn't be. Also it will run on real PDs and give us, well, a real e2e test.

Oops, you did that already. Thanks!

ernestas2k commented:
Updated the PR with requested changes. I appreciate the comments and quick feedback 👍

mattcary commented on pkg/common/utils.go:
```diff
@@ -161,15 +161,15 @@ func ConvertLabelsStringToMap(labels string) (map[string]string, error) {
 	}
 
-	regexKey, _ := regexp.Compile("^[a-z][-_a-z0-9]{0,62}$")
+	regexKey, _ := regexp.Compile("^\\p{Ll}[\\p{Ll}0-9_-]{0,62}$")
```
I'd prefer using a raw string (backtics) to avoid the doubled backslashes.

ernestas2k commented:
Done. Let me know if you find any more things to address.

saad-ali commented:
@ernestas2k could you also include the tags introduced in https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/570 (i.e. `"kubernetes.io/created-for/pv/name": "pvc-32af883a-d7cc-11ea-86f9-42010a800053"`, `"kubernetes.io/created-for/pvc/name": "pvc-demo"`, `"kubernetes.io/created-for/pvc/namespace": "default"`, and `"storage.gke.io/created-by": "pd.csi.storage.gke.io"`) in this PR to add as labels as well?

ernestas2k commented:
> #570

Sure, let's get answers to few questions first:
- Google only supports alphanumerical chars + "-" + "_" in keys and values, the keys in the PR contain more than that.
- Are they non-overridable, meaning we add them to the Labels map[string]string last so users won't be able to override them?

One more thing, the PR has definitions of Tag on disk parameters. This make me a bit worried they're interrelated and (possibly) should even be the same thing as Label in this PR. Thoughts?

ernestas2k commented:
> @ernestas2k could you also include the tags introduced in #570 (i.e. `"kubernetes.io/created-for/pv/name": "pvc-32af883a-d7cc-11ea-86f9-42010a800053"`, `"kubernetes.io/created-for/pvc/name": "pvc-demo"`, `"kubernetes.io/created-for/pvc/namespace": "default"`, and `"storage.gke.io/created-by": "pd.csi.storage.gke.io"`) in this PR to add as labels as well?

Sure, let's get answers to few questions first:

Google only supports alphanumerical chars + "-" + "_" in keys and values, the keys in the PR contain more than that.
Are they non-overridable, meaning we add them to the Labels map[string]string last so users won't be able to override them?
One more thing, the PR has definitions of Tag on disk parameters. This make me a bit worried they're interrelated and (possibly) should even be the same thing as Label in this PR. Thoughts?

saad-ali commented:
Good questions.

> Sure, let's get answers to few questions first:
> 
> - Google only supports alphanumerical chars + "-" + "_" in keys and values, the keys in the PR contain more than that.

Looks like the values shouldn't be an issue, but the keys will be.

How about we have a function that "sanitizes" key and value strings: specifically this sanitize method could and replaces non-allowed characters with `_` (underscore) (e.g. "kubernetes.io/created-for/pv/name: pvc-32af883a-d7cc-11ea-86f9-42010a800053" would become "kubernetes.io_created-for_pv_name": "pvc-32af883a-d7cc-11ea-86f9-42010a800053"), it could trim strings longer than max allowed, etc.).

> - Are they non-overridable, meaning we add them to the Labels map[string]string last so users won't be able to override them?

There should be no good reason for user to want to override them so, yes let's make it so user won't be able to override.
 
> One more thing, the PR has definitions of Tag on disk parameters. This make me a bit worried they're interrelated and (possibly) should even be the same thing as Label in this PR. Thoughts?

The only reason we added those "tags" in "description" is for feature parity with the legacy in-tree GCE PD volume plugin (the old in-tree GCE PD volume plugin used put tags as JSON in the description field" since labels didn't exist at that time). Labels are the correct way of handling these so we want labels, but the description tags are there in case some old component still depends on them.

saikat-royc commented:
What would be the behavior of validating exisiting disk? do we comapre labels too? 
https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-cloud-provider/compute/gce-compute.go#L268
func (cloud *CloudProvider) ValidateExistingDisk
@saad-ali @mattcary

saikat-royc commented on pkg/common/utils_test.go:
```diff
@@ -295,3 +295,155 @@ func TestKeyToVolumeID(t *testing.T) {
+				expectedError: true,
+			},
+			{
+				name:          "label key can have up to 63 characters",
```
nit: Can we group the failure cases and success cases for better readability

saikat-royc commented on pkg/common/utils_test.go:
```diff
@@ -295,3 +295,155 @@ func TestKeyToVolumeID(t *testing.T) {
+				name:          "label value can have up to 63 characters",
+				labels:        "abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghij123=v",
+				expectedError: false,
+			},
```
Add some more positive test cases for keys and values which include combination of '-' and '_'. e.g blah-_

saikat-royc commented:
> Good questions.
> 
> > Sure, let's get answers to few questions first:
> > 
> > * Google only supports alphanumerical chars + "-" + "_" in keys and values, the keys in the PR contain more than that.
> 
> Looks like the values shouldn't be an issue, but the keys will be.
> 
> How about we have a function that "sanitizes" key and value strings: specifically this sanitize method could and replaces non-allowed characters with `_` (underscore) (e.g. "kubernetes.io/created-for/pv/name: pvc-32af883a-d7cc-11ea-86f9-42010a800053" would become "kubernetes.io_created-for_pv_name": "pvc-32af883a-d7cc-11ea-86f9-42010a800053"), it could trim strings longer than max allowed, etc.).
> 
> > * Are they non-overridable, meaning we add them to the Labels map[string]string last so users won't be able to override them?
> 
> There should be no good reason for user to want to override them so, yes let's make it so user won't be able to override.
> 
> > One more thing, the PR has definitions of Tag on disk parameters. This make me a bit worried they're interrelated and (possibly) should even be the same thing as Label in this PR. Thoughts?
> 
> The only reason we added those "tags" in "description" is for feature parity with the legacy in-tree GCE PD volume plugin (the old in-tree GCE PD volume plugin used put tags as JSON in the description field" since labels didn't exist at that time). Labels are the correct way of handling these so we want labels, but the description tags are there in case some old component still depends on them.

^\p{Ll}[\p{Ll}0-9_-]{0,62}$ does not accept '.' e.g. kubernetes.io_created-for_pv_name

saikat-royc commented on pkg/common/utils_test.go:
```diff
@@ -295,3 +295,155 @@ func TestKeyToVolumeID(t *testing.T) {
+			{
+				name:          "label key may not have over 63 characters",
+				labels:        "abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghij1234=v",
+				expectedError: true,
```
and couple of more invalid cases can be added for . and /
			{
				name:          "label key cannot contain . and /",
				labels:        "kubernetes.io/created-for/pvc/namespace=v",
				expectedError: true,
			},
			{
				name:          "label value cannot contain . and /",
				labels:        "kubernetes_io_created-for_pvc_namespace=v./",
				expectedError: true,
			},

mattcary commented:
> What would be the behavior of validating exisiting disk in CreateVolume calls? do we comapre labels too?

AFAICT we don't have the PV plumbed through so it would be hard to check even if we wanted to. We also would only want to check the labels in the case where the PV was dynamically provisioned, information which we also don't seem to have.

Also, if the user were to delete the labels, and we enforced validation, it would be very hard to recreate them correctly. Since AFAIU the labels are used for informational purposes (accounting etc) it would be better to have a tool to recreate them appropriately rather than risk breaking storage usage if they get out of sync.

mattcary commented:
@ernestas2k any updates---are you waiting on anything?

msau42 commented:
Since the request for supporting the k8s metadata came in late, would it be simpler to split up into two PRs and treat the k8s metadata separately?

ernestas2k commented:
I'm in melancholic autumn mood right now, but since you pinged me I'll address your points this week 👍 . Unless you reconsider and split the task. Until then I'll address all of the points added.

mattcary commented:
> I'm in melancholic autumn mood right now, but since you pinged me I'll address your points this week 👍 . Unless you reconsider and split the task. Until then I'll address all of the points added.

Thanks! Summer is still in full swing here in Seattle and I still have a month before the autumn rains set in :-)

I think if you just make the test improvements Saikat mentioned, I'm okay with leaving the tag addition to a future CL. Let me know if you agree and I'll make in issue to track the tags (we're working on the filestore driver and will be adding labels in the same way).

ernestas2k commented:
Just updated the PR, addressed comments on test cases. 

Regarding kubernetes tags and sanitizing keys/values. Pretty straightforward change, but I would be surprized when strings that I've passed are accepted and **silently** converted. Take parameter `my_company/my_object_id=something/something-1-2-3` - conversion would change my key and value in some way that makes me not be able to query the disk anymore. At least, it would be a big surprise.

At most, I'd go with sanitizing the predefined keys of tag* constants (for example `tagKeyCreatedForClaimNamespace = "kubernetes.io/created-for/pvc/namespace"`). Tag values are also okay if it's guaranteed their values are always 'compliant' with google rules or the actual values aren't really that important. 

What are your thoughts?

mattcary commented:
> Just updated the PR, addressed comments on test cases.
> 
> Regarding kubernetes tags and sanitizing keys/values. Pretty straightforward change, but I would be surprized when strings that I've passed are accepted and **silently** converted. Take parameter `my_company/my_object_id=something/something-1-2-3` - conversion would change my key and value in some way that makes me not be able to query the disk anymore. At least, it would be a big surprise.
> 
> At most, I'd go with sanitizing the predefined keys of tag* constants (for example `tagKeyCreatedForClaimNamespace = "kubernetes.io/created-for/pvc/namespace"`). Tag values are also okay if it's guaranteed their values are always 'compliant' with google rules or the actual values aren't really that important.
> 
> What are your thoughts?

But since these labels are actually put on the PDs, what happens if an invalid character is passed through? Will PD creation fail or will something be silently dropped?

I think the common case would be that users would prefer to have labels actually created rather than having PD creation fail (which will occur some time after the SC creation and so be hard to debug).

We could also think about a validating webhook as is currently being done for snapshotting.

mattcary commented:
/remove-lifecycle rotten
/lifecycle frozen

StokeHead commented:
Just wanted to check if this issue is still being worked on?
I followed #693 and #570 , however both dont seem to exactly implement what i would need.

We are applying different actions on GCE PDs according to the originating StorageClass. However with the description field implemented in #570 i am only able to see the PVC/PV name. Without querying the cluster itself its therefore not possible to get the originating StorageClass. (Or am i missing something?)
Having the possibility to add custom labels, or extending the description field with the originating storage class should either be of help in my use case.

mattcary commented:
I'll take over this PR and get it in. Watch #340 for the status.

/close
